## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [107]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [108]:
# View the column names in the data

df.columns




Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [109]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [110]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.count()



first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [111]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df['category'].unique()

array(['decor', 'consumables', 'software', 'furniture', 'electronics'],
      dtype=object)

In [112]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.nunique()

first            360
last             477
job              498
phone            999
email            999
client_id        999
order_id       10000
order_date       150
order_week        23
order_year         2
item_id          500
category           5
subcategory       18
unit_price       490
unit_cost        487
unit_weight      393
qty             2911
line_number       10
dtype: int64

In [113]:
df['order_year'].unique()

array([2023, 2022], dtype=int64)

In [114]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [115]:
# For the category with the most entries,
# which subcategory had the most entries?

most_entries_df = df.loc[df['category'] =='consumables']
most_entries_df['subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [116]:
# Which five clients had the most entries in the data?

df['client_id'].value_counts().head()



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [117]:
# Store the client ids of those top 5 clients in a list.

#create DF for top 5 clients
client_ID_df = df['client_id'].value_counts().head().reset_index()

#rename DF columns
client_ID_df.columns = ['client_id', 'count']

#pass client ID column into list
client_ID_list = client_ID_df['client_id'].tolist()
client_ID_list


[33615, 66037, 46820, 38378, 24741]

In [118]:
# How many total units (the qty column) did the
# client with the most entries order order?

top_client_df = df.loc[df['client_id'] == 33615]
top_client_df['qty'].sum()   



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [119]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

#create a variable that calculates subtotal
subtotal_price = df['unit_price'] * df['qty']

#add subtotal to the dataframe
df['subtotal_price'] = subtotal_price
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  subtotal_price  
0     1096.80     762.71         7.50  105            1       115164.00  
1       24.95      15.09         1.49   21            0          523.95  
2       13.52       7.86         1.68   39            6          527.28  
3       36.42      24.85         1.23   29            3         1056.18  
4      195.10     108.17        46.43   20            1         3902.00

In [120]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

#create column for total_weight and calculate amount
total_weight = df['unit_weight'] * df['qty']

df['total_weight'] = total_weight

#add column and define function to return correct shipping weight based on weight
def shipping_rate_per_lb(weight):
    return 7 if weight > 50 else 10

#apply the function

df['shipping_rate_per_lb'] = df['total_weight'].apply(shipping_rate_per_lb)

#create column for shipping price
shippping_price = df['total_weight'] *df['shipping_rate_per_lb']

df['shipping_price'] = shippping_price
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  subtotal_price  total_weight  \
0         7.50  105            1       115164.00        787.50   
1         1.49   21            0          523.95         31.29   
2         1.68   39            6          527.28         65.52   
3         1.23   29            3         1056.18         35.67   
4        46.43   20            1         3902.00        928.60   

   shipping_rate_per_lb  shipping_price  
0                     7         5512.50  
1                    10          312.90  
2                     7          458.64  
3                    10          356.70  
4                     7         6500.20  

[5 rows x 22 columns]

In [121]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
# note that the final totals provided that we check against include shipping costs in the amount that is taxed.

total_price = round(((df['total_weight'] * df['shipping_rate_per_lb'] + df['subtotal_price']) * 1.0925),2)

df['total_price'] = total_price
df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  subtotal_price  total_weight  shipping_rate_per_lb  \
0            1       115164.00        787.50                     7   
1            0          523.95         31.29                    10   
2            6          527.28         65.52                     7   
3            3         1056.18         35.67                    10   
4            1         3902.00        928.60                     7   

   shipping_price  total_price  
0         5512.50    131839.08  
1          312.90       914.26  
2          458.64      1077.12  
3          356.70      1543.57  
4         6500.20     11364.40  

[5 rows x 23 columns]

In [122]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

total_cost = round((df['unit_cost'] * df['qty']) + (df['shipping_price']),2)

df['total_cost'] = total_cost
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   subtotal_price  total_weight  shipping_rate_per_lb  shipping_price  \
0       115164.00        787.50                     7         5512.50   
1          523.95         31.29                    10          312.90   
2          527.28         65.52                     7          458.64   
3         1056.18         35.67                    10          356.70   
4         3902.00        928.60                     7         6500.20   

   total_price  total_cost  
0    131839.08    85597.05  
1       914.26      629.79  
2      1077.12      765.18  
3      1543.57     1077.35  
4     11364.40     8663.60  

[5 rows x 24 columns]

In [123]:
# Create a column for the profit of
# each line using line cost and line price

profit = round((df['total_price'] - df['total_cost']),2)

df['line_profit'] = profit
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  subtotal_price  \
0          17        2023  ...        7.50  105           1       115164.00   
1          20        2023  ...        1.49   21           0          523.95   
2           4        2023  ...        1.68   39           6          527.28   
3          17        2023  ...        1.23   29           3         1056.18   
4          21        2023  ...       46.43   20           1         3902.00   

   total_weight  shipping_rate_per_lb  shipping_price  total_price  \
0        787.50                     7         5512.50    131839.08   
1         31.29                    10          312.90       914.26   
2         65.52                     7          458.64      1077.12   
3         35.67                    10          356.70      1543.57   
4        928.60                     7         6500.20     11364.40   

   total_cost  line_profit  
0    85597.05     46242.03  
1      629.79       284.47  
2      765.18       311.94  
3     1077.35       466.22  
4     8663.60      2700.80  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [124]:
# Check your work using the totals above

#create dataframe with only order_id and total_price columns for the order id's we will check. Create dataframes for each order_id.
#include only the order id and total price columns
#set index to order_id
checks_df_071 = df.loc[df['order_id'] == 2742071, ['order_id', 'total_price']].copy().set_index('order_id')
checks_df_913 = df.loc[df['order_id'] == 2173913, ['order_id', 'total_price']].copy().set_index('order_id')
checks_df_929 = df.loc[df['order_id'] == 6128929, ['order_id', 'total_price']].copy().set_index('order_id')

checks_df_913.head()


total_price
order_id             
2173913       9856.86
2173913        538.03
2173913       7092.74
2173913       5076.03
2173913       9640.08

In [125]:
#check work cont'd
#calculate totals for each order number

totals_x071=checks_df_071['total_price'].sum()
totals_x913=checks_df_913['total_price'].sum()
totals_x929=checks_df_929['total_price'].sum()

formatted071 = f"{totals_x071:,.2f}"
formatted913 = f"{totals_x913:,.2f}"
formatted929 = f"{totals_x929:,.2f}"

print(f"Order ID 2742071: {formatted071}")
print(f"Order ID 2173913: {formatted913}")
print(f"Order ID 6128929: {formatted929}")


Order ID 2742071: 152,811.89
Order ID 2173913: 162,388.71
Order ID 6128929: 923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [126]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

#create top 5 data frame referring to client ID's in previously created top 5 list
top_5_totals_df = df[df['client_id'].isin(client_ID_list)]
#top_5_totals_df

#calculate totals for each client id
client_id=top_5_totals_df.index.unique()
client_id

Index([    5,    12,   137,   214,   216,   357,   416,   459,   470,   520,
       ...
       54049, 54279, 54301, 54361, 54510, 54535, 54536, 54579, 54589, 54613],
      dtype='int64', length=1054)

In [127]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

#create top 5 data frame referring to client ID's in previously created top 5 list
top_5_totals_df = df[df['client_id'].isin(client_ID_list)]
#top_5_totals_df

#calculate totals for each client id

def client_totals(client_id, dataframe):
    client_tots_df = dataframe[dataframe['client_id'] == client_id]
    return round(client_tots_df['total_price'].sum(),2)


#create a list of dictionaries
client_tots_dict_list = [{'client_id': client_id, 'total_price': client_totals(client_id, top_5_totals_df)} for client_id in client_ID_list]

client_tots_df = pd.DataFrame(client_tots_dict_list)
client_tots_df=client_tots_df.reset_index(drop=True).set_index('client_id')
client_tots_df

total_price
client_id             
33615       8377308.52
66037      10259514.79
46820       9743794.36
38378      12906550.87
24741      82268892.02

In [128]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.


def client_all_totals(client_id, dataframe):
    client_id_tots_df = dataframe[dataframe['client_id'] == client_id]
    qty_tots = client_id_tots_df['qty'].sum()
    price_tots = client_id_tots_df['total_price'].sum()
    profit_tots = client_id_tots_df['line_profit'].sum()
    shipping_price_tots = client_id_tots_df['shipping_price'].sum()
    line_cost_tots = client_id_tots_df['total_cost'].sum()
    return {'client_id': client_id, 'qty': qty_tots, 'shipping_price': shipping_price_tots, 'total_price': price_tots,  'total_cost': line_cost_tots, 'total_profit': profit_tots, }


#create a list of dictionaries
all_tots = [client_all_totals(client_id,top_5_totals_df) for client_id in client_ID_list]
all_tots

#convert list of dictionaries into dataframe
all_tots_df = pd.DataFrame(all_tots)

#sort by total profit
all_tots_df_sorted = all_tots_df.sort_values('total_profit', ascending = False)
all_tots_df_sorted


client_id     qty  shipping_price  total_price   total_cost  total_profit
4      24741  239862      5126448.37  82268892.02  45688899.71   36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98    3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98    3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64    2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91    2201994.61

In [129]:
#check data types
all_tots_df_sorted.dtypes

client_id           int64
qty                 int64
shipping_price    float64
total_price       float64
total_cost        float64
total_profit      float64
dtype: object

In [130]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

#rename columns
renamed_df = all_tots_df_sorted.rename(columns = {'client_id': 'Client ID',
                                           'qty':  'Units',
                                           'shipping_price': 'Shipping',
                                           'total_price': 'Total Revenue',
                                           'total_cost': 'Total Cost',
                                           'total_profit': 'Total Profit'
                                           })

#format data to millions of dollars
def to_millions(row):
    return row / 1000000

#create list of columns to convert
columns_to_convert_list = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']

#convert $$ columns to millions, and add '$' and 'M'

renamed_df[columns_to_convert_list] = renamed_df[columns_to_convert_list].apply(lambda x: x.apply(lambda y: '${:.2f}M'.format(y / 1000000)))
renamed_df



Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M